proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -304,15 +451,196 @@ elimation has completed.
 
 ### Validations in Sema
 
-TODO
+#### All the values should be legal.
+
+Most values like ShaderVisibility/ParameterType are covered by syntactical checks in Sema.
+Only list special rule here.
+
+- For DESCRIPTOR_RANGE_FLAGS on a Sampler, only the following values are valid
+  - 0
+  - DESCRIPTORS_VOLATILE
+  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS
+
+- For DESCRIPTOR_RANGE_FLAGS on a CBV/SRV/UAV, only the following values are
+   valid
+  - 0
+  - DESCRIPTORS_VOLATILE
+  - DATA_VOLATILE
+  - DATA_STATIC
+  - DATA_STATIC_WHILE_SET_AT_EXECUTE
+  - DESCRIPTORS_VOLATILE | DATA_VOLATILE
+  - DESCRIPTORS_VOLATILE | DATA_STATIC_WHILE_SET_AT_EXECUTE
+  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS
+  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_VOLATILE
+  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC
+  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC_WHILE_SET_AT_EXECUTE
+
+- StaticSampler
+  - Max/MinLOD cannot be NaN.
+  - MaxAnisotropy cannot exceed 16.
+  - MipLODBias must be within range of [-16, 15.99].
+
+- Register Space
+  -The range 0xFFFFFFF0 to 0xFFFFFFFF is reserved.
+  ```
+  "CBV(b0, space=4294967295)" is invalid due to the use of reserved space 0xFFFFFFFF.
+  ```
+
+- Resource ranges must not overlap.
+  ```
+  "CBV(b2), DescriptorTable(CBV(b0, numDescriptors=5))" will result in an error
+  due to overlapping at b2.
+  ```
 
 ### Metadata Schema
 
 TODO
 
-### Validations during Codegen
+### Validations during DXIL generation
+  [Validations In Sema](#validations-in-sema)
+#### All the things validated in Sema.
+  In DXIL generation, values like ShaderVisibility/ParameterType need to be
+  checked to make sure they are in correct range.
+
+- Valid values for ShaderVisibility
+  - SHADER_VISIBILITY_ALL
+  - SHADER_VISIBILITY_VERTEX
+  - SHADER_VISIBILITY_HULL
+  - SHADER_VISIBILITY_DOMAIN
+  - SHADER_VISIBILITY_GEOMETRY
+  - SHADER_VISIBILITY_PIXEL
+  - SHADER_VISIBILITY_AMPLIFICATION
+  - SHADER_VISIBILITY_MESH
+
+- Valid values for ParameterType
+  - CBV
+  - SRV
+  - UAV
+  - DescriptorTable
+  - Constants32Bit
```

Comment:
In the metadata we have entries with names like:

* RootFlags
* RootCBV
* StaticSampler
* DescriptorTable

Maybe this'll be something that needs revisiting once the metadata format is fleshed out,

Reply:
The ParameterType here is only for the Parameter part of the root signature.
For the entries in metatdata, they're for the whole root signature, not the same thing.
We could visit metadata entries when fill the metadata section.
The parameter type here should be correct.

Reply:
At this point, I think we're validating that the values provided by the metadata are valid - so that if errors are encountered we can refer to them in terms of the metadata.

What, specifically, are you envisioning "ParameterType" is actually validating here? If it is the value of D3D12_ROOT_PARAMETER::ParameterType, then this is something that is generated by codegen - we shouldn't need to do a second pass of validation on the output of codegen. What we need to validate is the _input_ to codegen, which is the metadata.

Reply:
I see. Remove ParameterType.

Reply:
I've added a note to #234 about ensuring that the validation section includes information about what needs to be validated for the metadata.